### ISSUE TITLE
Request for guidance with a code splitting pattern

### ISSUE THREAD

**Issue by arggh** [VeliovGroup/flow-router#63]
Hello and thank you for taking care of Flow Router! I'm looking forward to taking advantage of all the nice new features implemented in `flow-router-extra`.

I only recently switched over from the abandoned Kadira-version, so apologies if I've missed something, but I couldn't find guidance in the docs for the following scenario. I think the changes in `this.render` vs. `BlazeLayout.render` might be throwing me off.

I'm trying to accomplish a pattern, where only my login page and few selected components are included in the initial bundle, and rest of the app is loaded with dynamic imports when accessing other routes.

My first implementation tried to achieve this by using a different layout for the remaining routes, called `app_layout`, that would import rest of the app when rendered.

```js
Template.app_layout.helpers({
  initializationComplete() {
    return Template.instance().initialized.get();
  }
});

Template.app_layout.onCreated(function () {
  this.initialized = new ReactiveVar(false);
  import('../initialize_app').then(() => this.initialized.set(true));
});
```

```js
// initialize_app.js
import './help';
import './main_layout';
import './header';
```

```handlebars
<template name="app">
    {{#if initializationComplete}}
      {{> yield}}
    {{/if}}
</template>
```

...and my route definitions are like this:

```js
FlowRouter.route('/', {
  name: 'landing',
  waitOn() {
    return import('../../client/modules/index');
  },
  action() {
    this.render('app', 'main_layout', {
      content: 'index',
      header: 'header',
      footer: 'help'
    });
  }
});
```

I was expecting the `app_layout` to import the commonly used `main_layout`, `header` and `help` templates, and once imported, render the `{{> yield }}` content.

However, this breaks with the following error thrown on client:

```
Exception from Tracker afterFlush function:
meteor.js?hash=857dafb4b9dff17e29ed8498a22ea5b1a3d6b41d:1059 Error: Can't render undefined
    at checkRenderContent (blaze.js?hash=51f4a3bdae106610ee48d8eff291f3628713d847:2285)
    at contentAsFunc (blaze.js?hash=51f4a3bdae106610ee48d8eff291f3628713d847:2328)
    at Object.Blaze.renderWithData (blaze.js?hash=51f4a3bdae106610ee48d8eff291f3628713d847:2401)
    at BlazeRenderer.materialize (renderer.js:341)
    at Blaze.TemplateInstance.<anonymous> (renderer.js:65)
    at blaze.js?hash=51f4a3bdae106610ee48d8eff291f3628713d847:3398
    at Function.Template._withTemplateInstanceFunc (blaze.js?hash=51f4a3bdae106610ee48d8eff291f3628713d847:3769)
    at fireCallbacks (blaze.js?hash=51f4a3bdae106610ee48d8eff291f3628713d847:3394)
    at Blaze.View.<anonymous> (blaze.js?hash=51f4a3bdae106610ee48d8eff291f3628713d847:3487)
    at blaze.js?hash=51f4a3bdae106610ee48d8eff291f3628713d847:1845
```

That had me thinking, maybe I should dynamically import the `main_layout` in the `waitOn` hook, and include the named slots in the template (as instructed in a few places), but then I would have to pepper all my route definitions with `waitOn` on the `main_layout`, since I never know which route the user chooses to access initially. Then again, I don't want to include `main_layout` in the initial bundle, as it uses a whole lotta other components that I'd rather not include in the login-only-initial-bundle.

Maybe I'm just confused from doing too much refactoring on a large codebase in the last 48hrs, but I'd gladly accept any help on how to get started with this sort of app loading pattern and flow-router-extra!

**Comment by arggh** [comment]
Another option I was thinking is to use the global `triggersEnter` hooks, excluding the login route, to check if the app has already been initialized, and dynamically import the required code if not. However, I suspect I can't really wait in the `triggersEnter`hook for the dynamic `import()` to resolve before rendering the route?

**Comment by arggh** [comment]
**TL;DR**

I guess the question boils down to: with `flow-router-extra`, how to best implement a common, shared wrapper for "logged in"-state, that dynamically imports all shared bits and pieces, when all routes are split-points for code splitting?

**Comment by coagmano** [comment]
I'm a little confused as to how the code snippets fit together, but if the issue is with `yield` and `this.render`, you could still use `BlazeLayout.render` like in classic FlowRouter?

**Comment by dr-dimitru** [comment]
Hello guys @arggh and @coagmano,

> I guess the question boils down to: with flow-router-extra, how to best implement a common, shared wrapper for "logged in"-state, that dynamically imports all shared bits and pieces, when all routes are split-points for code splitting?

You don't have to split all app apart, it might be a good idea to keep some of the small and reusable templates in the core bundle. Take a look at this [demo application](https://github.com/VeliovGroup/Meteor-Files-Demos/blob/master/demo/imports/lib/routes.js#L26).

For example, we use separate layout for pages accessible only to authorized users, if user authorized we conditionally show `{{> yield}}` template. When user is unauthorized we render `{{> loginForm}}` template. Something like that:

```jade
template(name="mainLayout")
  if currentUser
    +yield
  else
    +loginForm
```

> I'm a little confused as to how the code snippets fit together, but if the issue is with yield and this.render, you could still use BlazeLayout.render like in classic FlowRouter?

 1. No need to switch to `this.render()`, you're going to be fine with `BlazeLayout`
 2. Only differences of `this.render()` from `BlazeLayout`:
    1. Consider `{{> yield}}` as main "region"
    2. `this.render()` [actually supports regions](https://github.com/VeliovGroup/flow-router/blob/master/docs/templating-with-regions.md) 
    3. `this.render()` is highly optimized in terms of rendering, we're using Blaze API directly to render DOM elements off the screen, and run many other optimized operations.

> Another option I was thinking is to use the global triggersEnter hooks, excluding the login route, to check if the app has already been initialized, and dynamically import the required code if not. However, I suspect I can't really wait in the triggersEnterhook for the dynamic import() to resolve before rendering the route?

Yes, it's simply blocking async, no magic

> Hello and thank you for taking care of Flow Router! I'm looking forward to taking advantage of all the nice new features implemented in flow-router-extra.

I'm glad you've found it to be useful

> However, this breaks with the following error thrown on client:

Well, make sure templates `'app', 'main_layout'` are loaded and available via `Template[nameOfTemplate]`.

**Comment by arggh** [comment]
Thanks for the input.

Still not quite what I was looking for.

The best I could do for now, was to create a `logged_in` template that encapsulates all the imports and logic needed. That template is then being used in each of my layouts. This way, they share whatever is needed for the logged in -state, without having to handle the same jazz in each route or layout separately.

Simplified:

```js
Template.logged_in.viewmodel({
  onCreated() {
    import('../../initialize_app').then(() => this.importsLoaded(true));
  },

  ready() {
    return this.importsLoaded() && this.authenticated();
  }
});
```
```handlebars
<template name="logged_in">
  {{#if ready}}
    {{> Template.contentBlock}}
  {{/if}}
</template>
```

...and in the layouts:

```handlebars
<template name="main_layout">
  {{#logged_in}}
    {{> connection_lost ''}}
    {{> announcements ''}}

    {{#app}}
      {{> header ''}}
      {{> yield }}
    {{/app}}

    {{> help ''}}
    {{> mobile_menu ''}}
  {{/logged_in}}
</template>
```

```handlebars
<template name="basic_layout">
  {{#logged_in}}
    {{> connection_lost ''}}
    {{> yield }}
  {{/logged_in}}
</template>
```

etc... except for the layouts used by signup and login, which don't need any of the stuff imported by `logged_in`.

> Well, make sure templates 'app', 'main_layout' are loaded and available via Template[nameOfTemplate]

This was actually quite likely caused by a bug in the Meteor `1.8.1-beta.20` release, not FlowRouter.

**Comment by dr-dimitru** [comment]
@arggh I'd do it differently (in terms of code styling)

> This was actually quite likely caused by a bug in the Meteor 1.8.1-beta.20 release, not FlowRouter.

Well, might be a good idea to stick with *stable* channel, and [fyi] we're do not provide support on v.next releases.

@arggh shall we close it as *solved*?

**Comment by arggh** [comment]
> I'd do it differently (in terms of code styling)

Please, do tell me how you would do it differently in terms of code styling and why? Usually style is about preference so it's a bit weird to comment on it (like camelCase vs. snake_case), so maybe you meant something else?

> Well, might be a good idea to stick with stable channel, and [fyi] we're do not provide support on v.next releases.

I'm getting some condescending vibes from you, but maybe that's just me?

`<offtopic>`
I, and many others, don't always stick with "stable channel", but instead opt to help out in testing beta releases. In this case, my testing of the Meteor `1.8.1-beta.20` discovered the bug I was talking about, allowed me to create a reproduction and thanks to that, the bug was fixed before `rc.0` release by the Meteor core team.

Testing Meteor packages with Meteor's beta- and release candidate -releases is also valuable and might give _You_, a package developer and maintainer, a nice heads-up to fix something that's going to be broken with the next official release of Meteor.

I think it's an integral part of open source and would personally accept this sort of assistance with gratitude.
`</offtopic>`

I don't _expect_ to receive any support for an open source project. I was _hoping_ someone would point me in the right direction and if not, then I'm merely opening a discussion on something that I think might be improved in `flow-router-extra`, either in terms of documentation or design.

I don't think I had the best possible experience defining my routes with `flow-router-extra` when aiming for a code-splitted app with multiple entry points, which of some share common code and some don't, so maybe there is room to improve (in docs or design)?

Feel free to close this if you don't think this conversation brings any value to your project!

**Comment by dr-dimitru** [comment]
> Please, do tell me how you would do it differently in terms of code styling and why? Usually style is about preference so it's a bit weird to comment on it (like camelCase vs. snake_case), so maybe you meant something else?

Please take a look at our [demo app](https://github.com/VeliovGroup/Meteor-Files-Demos/tree/master/demo):

 1. [`router.js`](https://github.com/VeliovGroup/Meteor-Files-Demos/blob/master/demo/imports/client/router/router.js)
 2. [`routes.js`](https://github.com/VeliovGroup/Meteor-Files-Demos/blob/master/demo/imports/lib/routes.js)
 3. [`_layout.jade`](https://github.com/VeliovGroup/Meteor-Files-Demos/blob/master/demo/imports/client/misc/_layout.jade)

> I'm getting some condescending vibes from you, but maybe that's just me?

No, I just don't think it is a great idea to work on unstable branch. Testing is great

> I don't think I had the best possible experience defining my routes with flow-router-extra when aiming for a code-splitted app with multiple entry points, which of some share common code and some don't, so maybe there is room to improve (in docs or design)?

We're always open to discussions, especially about code design, as we inherited and extended API of original Flow Router we had not much of the room to play around, trying to follow original design.

**Comment by arggh** [comment]
> Please take a look at our demo app:

I did, and to make it more obvious, here I extracted all relevant bits of code from the linked files:

```js
// 1. router.js
// nothing relevant
```

```js
// 2. routes.js
FlowRouter.route('/', {
  name: 'index',
  action() {
    this.render('_layout', 'index');
  },
  waitOn() {
    if (Meteor.isClient) {
      return import('/imports/client/index/index.js');
    }
  }
});

FlowRouter.route('/login', {
  name: 'login',
  action() {
    this.render('_layout', 'login');
  },
  waitOn() {
    return import('/imports/client/user-account/login.js');
  }
});

FlowRouter.route('/:_id', {
  name: 'file',
  action(params, queryParams, file) {
    this.render('_layout', 'file', {
      file: file
    });
  },
  waitOn(params) {
    return import('/imports/client/file/file.js');
  }
});
```

```jade
// _layout.jade
template(name="_layout")
 +yield
```

Unless I'm mistaken, this demonstrates the basic usage of Flow Router with dynamic imports. It's simply dynamically importing a template, which is then rendered using the layout that all the routes share.

Nowhere does the demo app address the problem of having some common code shared between some, but not all, routes. Maybe I'm expressing myself poorly, so I'll try once more with an imaginary situation:

_I would like to have..._

- A login&signup -page that is really lightweight and doesn't include almost anything in common with the rest of the app
- 2 special routes, which load a tiny portion of the app, but stay lightweight similar to the login page
- 10 different layouts
- 256 other routes, which combined use all 10 different layouts depending on route, and all require a big portion of common code to be loaded. Not just a few helper functions that you could include in the signup route, but 6Mb of code shipped down the wire
- Any of those 256 routes can be an "entry" route to the app

Now, with `flow-router-extra`, how do I best share that common code between those 256 routes, in a way that...

- doesn't force me to load the same 6Mb of code on the login&signup -page and the 2 special routes
- doesn't force me to write `import './common_code';` 256 times in each entry-point file
- doesn't force me to write the dynamic import logic in equal fashion in all 10 layouts

**?**

The two solutions that I came up with, was...

1. Use two layouts:
```js
FlowRouter.route('/', {
  waitOn() {
    return import('./index');
  },
  action() {
    this.render('logged_in_layout', 'actual_layout', {
      tmpl: 'index'
    });
  }
});
```
...and in `logged_in_layout`, load the common code shared between the 256 routes via dynamic `import()` before allowing `{{> yield}}` to be rendered. This option forces me to use the named "slots" for the actual content in `actual_layout`, and to use the more complicated render calls on all 256 routes.

2. Use a `logged_in` helper template in the 10 layouts, that dynamically imports the common shared code between the 256 routes:

```handlebars
<template name="some_layout">
  {{#logged_in}}
    {{> yield }}
  {{/logged_in}}
</template>
```

```js
FlowRouter.route('/', {
  waitOn() {
    return import('./index');
  },
  action() {
    this.render('some_layout', 'index');
  }
});
```

`logged_in` only renders `{{> yield}}` once the common code shared by 256 routes is imported dynamically and it's used in all the layouts that are used for the "logged in" -state routes. This doesn't have the downsides of option 1.

3. This had me thinking, maybe there is a better way to handle this situation, an option 3, but I couldn't find it in the docs... which lead to me asking about it here.

**Comment by dr-dimitru** [comment]
> doesn't force me to write import './common_code'; 256 times in each entry-point file

This is the right approach, like most of the JS files in Meteor project starts with `import { Meteor } from 'meteor/meteor';` all JS files with view-controller logic should have a line `import './common_code';`

Here is more complex example:

```jade
template(name="_layout")
  +sidebar
  .container
    //- check if user visiting "protected" (behind login) route
    if compare currentRouteName 'is' 'index|signup|signin'
      //- User visiting public page, render its template:
      +yield
    else if currentUser
      //- User logged-in and visiting "protected" route:
      if isVerified
        if currentUser.profile.plan
          //- if user logged in, verified, and has selected billing plan show route-defined template
          +yield
        else
          //- if user logged in, verified nut has NOT selected billing plan yet — show billing template
      else
        //- show verification template when user logged in but not verified
    else
      // User NOT logged-in and visiting "protected" route — show sign-in form:
      +signin

  if isMobile
    footer.gap-footer &nbsp;
    .mobile-nav: ul
      //- Hide footer and show mobile navigation
  else
    footer.page-footer(itemscope itemtype="http://schema.org/Organization")
       //- footer
```

`routes.js` example:

```js
FlowRouter.route('/', {
  name: 'index',
  whileWaiting: loader.start,
  endWaiting: loader.stop
  waitOn() {
    return import('/imports/client/index.js');
  },
  action() {
    this.render('_layout', 'index');
  }
});

FlowRouter.route('/login', {
  name: 'signin',
  title: 'Sign In',
  action() {
    if (Meteor.user()) {
      FlowRouter.go('dashboard');
    } else {
      this.render('_layout', 'signin');
    }
  },
  waitOn() {
    return import('/imports/client/sign-in/sign-in.js');
  },
  whileWaiting: loader.start,
  endWaiting: loader.stop
});
```

**Comment by coagmano** [comment]
Meteor will de-dupe dynamic imports and use the client side cache for modules being requested more than once, so sticking it in `waitOn` is a viable strategy.

Even better, you can make a group with the `waitOn` defined and it will apply to all routes added to that group:
```js
const loggedInGroup = FlowRouter.group({
  name: 'logged-in',
  triggersEnter() {
    // redirect if not logged in etc.
  },
  waitOn: () => import('./common_code')
});
// then define the routes that depend on common_code:
loggedInGroup.route('/example', {
  // ...
});
```

Only issue I can think of is that triggersEnter runs after waitOn so redirection also waits

**Comment by dr-dimitru** [comment]
@arggh yes, I forgot to tell you about *group* it's a great way to achieve your case. Thank @coagmano for your example

**Comment by arggh** [comment]
> This is the right approach, like most of the JS files in Meteor project starts with import { Meteor } from 'meteor/meteor'; all JS files with view-controller logic should have a line import './common_code';

I think you are right, but also that it's not _wrong_ to import the common code in (for example) the wrapping templates, especially if it spares you from writing those 256 lines, and if it's not something you import for consuming, rather something that has side-effects, sets up some global state, imports Blaze templates, inits subscriptions etc...

> Even better, you can make a group with the waitOn defined and it will apply to all routes added to that group

Defining `waitOn` in the group is something I tried, but it's also somewhat incompatible with views that use the same template but multiple routes (using the path for storing state of the app) and use `whileWaiting` - Whenever the route changes, `waitOn`of the group will kick in, wait for the promise to resolve (even if the module is cached) and for a  ~hundred milliseconds, the template defined in `whileWaiting` will be rendered, before rendering the same template again.

> Only issue I can think of is that triggersEnter runs after waitOn so redirection also waits

This was something I had not considered, thanks!

1. If I had to imagine an API that I would have used if it existed, maybe it would have been like this:

```
FlowRouter.hooks.waitOn([import('./common_code')], { except: ['login'], once: true });
```

2. As a suggestion for the docs, maybe some of the strategies explained here could be added?

**Comment by dr-dimitru** [comment]
@arggh you know there is no need to render a template in `whileWaiting`, what we do is we waiting for 200ms if app is still loading something — show overlay loader, if not proceed to `.action()`.

> FlowRouter.hooks.waitOn([import('./common_code')], { except: ['login'], once: true });

I believe you can check name of the route and import `./common_code` conditionally in waitOn hook

> As a suggestion for the docs, maybe some of the strategies explained here could be added?

PRs and contributions are very welcome.

**Comment by dr-dimitru** [comment]
Hello guys 👋

Shall we close this one? It's definitely beyond package bounds, all hooks are explained with suggestions and usage examples. There is a lot of ways to accomplish and one of them is using hooks of `flow-router`.

@arggh feel free to close it.

**Comment by arggh** [comment]
Sure. Please re-open if you actually would like a PR about this (I got the feeling that you didn't see the value)?

And in that case I would like some opinions on which strategy (or all?) to document.
